File: class1_ROS_Basic/act_sercli_example_pkg/src/simple_action_server.py
#! /usr/bin/env python
import rospy
import actionlib
from custom_messages.msg import CounterAction, CounterGoal, CounterResult, CounterFeedback

class CounterClass(object):
    
    # create messages that are used to publish feedback/result
    _feedback = CounterFeedback()
    _result   = CounterResult()

    def __init__(self):
        # creates the action server
        self._as = actionlib.SimpleActionServer("counter_server", CounterAction, self.goal_callback, False)
        self._as.start()
        
    def goal_callback(self, goal):
        # this callback is called when the action server is called.
        # this is the function that computes the Fibonacci sequence
        # and returns the sequence to the node that called the action server
        
        # helper variables
        r = rospy.Rate(1)
        success = True
        
        # append the seeds for the fibonacci sequence
        self._feedback.percentage = 0
        # publish info to the console for the user
        rospy.loginfo('"Counter Server"')
        
        # starts calculating the Fibonacci sequence
        maxnumber = goal.max_number
        duration = goal.duration
        
        cur_count = 0
        
        r = rospy.Rate(duration)
        for i in range(0, maxnumber + 1):
        
        # check that preempt (cancelation) has not been requested by the action client
            if self._as.is_preempt_requested():
                rospy.loginfo('The goal has been cancelled/preempted')
                # the following line, sets the client in preempted state (goal cancelled)
                self._as.set_preempted()
                success = False
                # we end the calculation of the Fibonacci sequence
                break
        
            # builds the next feedback msg to be sent
            
            self._feedback.percentage = cur_count/maxnumber
            cur_count += 1
            
            # publish the feedback
            self._as.publish_feedback(self._feedback)
            # the sequence is computed at 1 Hz frequency
            r.sleep()
        
        # at this point, either the goal has been achieved (success==true)
        # or the client preempted the goal (success==false)
        # If success, then we publish the final result
        # If not success, we do not publish anything in the result
        if success:
            self._result.count = maxnumber
            rospy.loginfo('Succeeded!! Count is %i' % self._result.count )
            self._as.set_succeeded(self._result)
        
if __name__ == '__main__':
    rospy.init_node('Counter_server')
    CounterClass()
    rospy.spin()

File: class1_ROS_Basic/act_sercli_example_pkg_c/src/simple_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <custom_messages/CounterAction.h>

class CounterClass
{
protected:

  ros::NodeHandle nh_;
  // NodeHandle instance must be created before this line. Otherwise strange error occurs.
  actionlib::SimpleActionServer<custom_messages::CounterAction> as_; 
  std::string action_name_;
  // create messages that are used to publish feedback and result
  custom_messages::CounterFeedback feedback_;
  custom_messages::CounterResult result_;

public:

  CounterClass(std::string name) :
    as_(nh_, name, boost::bind(&CounterClass::executeCB, this, _1), false),
    action_name_(name)
  {
    as_.start();
  }

  ~CounterClass(void)
  {
  }

  void executeCB(const custom_messages::CounterGoalConstPtr &goal)
  {
    // helper variables
    int cur_count = 0;
    int maxnumber = 0;
    int duration = 0;

    maxnumber = goal->max_number;
    duration = goal->duration;
    ROS_INFO("%d: maxnumber", maxnumber);
    ROS_INFO("%d: duration", duration);
    bool success = true;

    ros::Rate r(duration);
    feedback_.percentage = 0;

    for(int i = 0; i < maxnumber +1 ; i++)
    {
      if (as_.isPreemptRequested() || !ros::ok())
      {
        ROS_INFO("%s: Preempted", action_name_.c_str());
        // set the action state to preempted
        as_.setPreempted();
        success = false;
        break;
      }
      feedback_.percentage = (float)(cur_count)/(float)(maxnumber);
      cur_count++;
      ROS_INFO("%d: cur_count", cur_count);
      as_.publishFeedback(feedback_);
      r.sleep();
    }
    
    ROS_INFO("Counter Server");
  
    if(success)
    {
      result_.count = maxnumber;
      ROS_INFO("%s: Succeeded", action_name_.c_str());
      as_.setSucceeded(result_);
    }
  }


};


int main(int argc, char** argv)
{
  ros::init(argc, argv, "counter_server");

  CounterClass counter("counter_server");
  ros::spin();

  return 0;
}